### PR TITLE
Removes calls to MeasureText() in chatmessage, approximates message's width and height

### LINKED ISSUE #55368: Runechat will not display the "U" character if it's near the end of a line
## Testmerges:
#54588: 'Bring back Kilostation v3.0' by Maurukas at commit 81337211e9
#55246: 'Filterrific!' by actioninja at commit 001061eaf3
#55285: 'Remove SM from delta station atmos and put it in engi' by Ghilker at commit b302d12381

## Reproduction:
This has been happening for a while now, but if the character U appears near the end of a line, it just won't appear.
Examples:

image
image

image
image

image
image

image
image

### PR DESCRIPTION
**Pull request by Archanial**
## About The Pull Request

Copy of https://github.com/BeeStation/BeeStation-Hornet/pull/6788 so I'd suggest waiting with merge until the first one is merged.
I'd also suggest testmerging this first.

This PR adds approximation of messages displayed width and height. To do that we calculate actual values when the first client logs in and cache them in a list. Then we iterate over characters checking their precalculated width or assuming the worst if the character is unusual.

closes: #55368

This is largely inspired by @Kylerace PR with other optimizations.
## Why It's Good For The Game

Makes the chatmessage init synchronous which prevents some unwanted behavior from happening and of course this is faster than calling something on client.

During initial testing by making 250 random chatmessages then testing MeasureText against my approximation there was average of +-1% difference in calculated values.
On top of that I've found that there is at least one example where my approximation is actually more accurate than the previous one.

The cache is now per client.

Every character that is not found with valid size will be added to a list and added to client's cache.

All the screens bellow come from TG:
image
image
image
image
image
image

## Changelog

:cl:
refactor: implemented better approximation of chatmessages height and width calculation
/:cl:

### CODE REVIEW

File: code/controllers/subsystem/runechat.dm
Comment (LemonInTheDark) [inline code/controllers/subsystem/runechat.dm:88]: why is this a list entry and not just a var on the subsystem

File: code/controllers/subsystem/runechat.dm
Comment (LemonInTheDark) [inline code/controllers/subsystem/runechat.dm:107]: surely there's a better way of accomplishing this? also what happens if the client lies to us?

File: code/datums/chatmessage.dm
Comment (LemonInTheDark) [inline code/datums/chatmessage.dm:50]: Where'd these go? were they just unused?

File: code/datums/chatmessage.dm
Comment (LemonInTheDark) [inline code/datums/chatmessage.dm:194]: why use ? here?

File: code/datums/chatmessage.dm
Comment (LemonInTheDark) [inline code/datums/chatmessage.dm:332]: does this behave poorly with unicode and friends then?

File: code/datums/chatmessage.dm
Comment (LemonInTheDark) [inline code/datums/chatmessage.dm:30]: you should wrap this in (), so someone sticking * 2 on the end of it gets what they expect

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
Some concerns, I'm mostly worried about the "bad client" case, and also kinda the overhead of that string iteration

File: code/datums/chatmessage.dm
Comment (Archanial) [reply code/datums/chatmessage.dm:50]: They were unused

File: code/datums/chatmessage.dm
Comment (Archanial) [reply code/datums/chatmessage.dm:192]: It's the same as as anything

File: code/datums/chatmessage.dm
Comment (Archanial) [reply code/datums/chatmessage.dm:332]: You shouldn't spam weird Unicode symbols anyway. If the symbol is bigger than biggest symbol (M) it may cut the message in some really edge cases.

File: code/datums/chatmessage.dm
Comment (Archanial) [reply code/datums/chatmessage.dm:346]: It's linear with access to list. It may look hot but you essentially perform two simple operations

File: code/datums/chatmessage.dm
Comment (Archanial) [reply code/datums/chatmessage.dm:163]: What do you mean?
Do you have different message mods here?

File: code/datums/chatmessage.dm
Comment (Archanial) [reply code/datums/chatmessage.dm:194]: I've had some issues with m being null before, not sure if that'll be case anymore actually.

File: code/controllers/subsystem/runechat.dm
Comment (Archanial) [reply code/controllers/subsystem/runechat.dm:88]: No reason in particular. Is there any difference though?

Comment (LemonInTheDark) [reply code/controllers/subsystem/runechat.dm:88]: It seems pointlessly unwieldy, and kinda uh, mixes the meaning of the list

File: code/datums/chatmessage.dm
Comment (LemonInTheDark) [reply code/datums/chatmessage.dm:332]: yeah but if people do are we gonna fuck them?

Comment (Kylerace) [reply code/datums/chatmessage.dm:332]: 😳

Comment (Mothblocks) [reply code/datums/chatmessage.dm:332]: Your code absolutely should be prepared for Unicode text, because not all Unicode text is emojis. People using accents, or even something like Chinese characters, are all things that the codebase should be prepared for.

File: code/controllers/subsystem/runechat.dm
Comment (Mothblocks) [inline code/controllers/subsystem/runechat.dm:100]: Better variable name

File: code/datums/chatmessage.dm
Comment (Archanial) [reply code/datums/chatmessage.dm:163]: Actually those are styles defined in skin.dmf

File: code/datums/chatmessage.dm
Comment (Archanial) [reply code/datums/chatmessage.dm:332]: 99% characters used in speak are not unicode characters. By assuming the largest pixel count (M which is about 10px) the only things that will cause actual issues are the weird super-long characters and spamming them is definitely malicious.
The worst that could happen is that our message will take one more line, which is small visual bug, that could happen already.

I guess if you could prepare some txt files with all the chars we want to track and read it and calculate our value based on those characters.

**Comment by MrDoomBringer** [comment]
Does this PR affect https://github.com/tgstation/tgstation/issues/55368 in any way?

File: code/datums/chatmessage.dm
Comment (ike709) [reply code/datums/chatmessage.dm:192]: faster to type

File: code/controllers/subsystem/runechat.dm
Comment (Archanial) [reply code/controllers/subsystem/runechat.dm]: Addressed

File: code/datums/chatmessage.dm
Comment (Archanial) [reply code/datums/chatmessage.dm]: addressed

File: code/controllers/subsystem/runechat.dm
Comment (Archanial) [reply code/controllers/subsystem/runechat.dm]: I've changed the comment

File: code/controllers/subsystem/runechat.dm
Comment (Archanial) [reply code/controllers/subsystem/runechat.dm]: Addressed

**Comment by Archanial** [comment]
> Does this PR affect #55368 in any way?

Yes. This is my second encounter where MeasureText() was less accurate.
It's working fine on after the changes.
image

File: code/datums/chatmessage.dm
Comment (Mothblocks) [reply code/datums/chatmessage.dm:339]: > The worst that could happen is that our message will take one more line

That's fine then. Unicode characters *extending* the size is fine, since at least you can still see everything.

File: code/controllers/subsystem/runechat.dm
Comment (LemonInTheDark) [reply code/controllers/subsystem/runechat.dm]: Why couldn't they? It'd be a matter of sending a false response for measure text calls, should be possible

File: code/datums/chatmessage.dm
Comment (LemonInTheDark) [reply code/datums/chatmessage.dm:200]: it's weird to as anything the loop and then null check anyway

File: code/datums/chatmessage.dm
Comment (LemonInTheDark) [reply code/datums/chatmessage.dm:339]: I would be interested in seeing what this does to accent characters, and also kinda russian

File: code/datums/chatmessage.dm
Comment (LemonInTheDark) [reply code/datums/chatmessage.dm]: should still cache SSrunechat.letters locally to improve lookup times

**Comment by MrDoomBringer** [comment]
Huge Dub

File: code/datums/chatmessage.dm
Comment (Kylerace) [reply code/datums/chatmessage.dm:339]: @Archanial a 100% safe solution is to do the approximation unless it scans a character that isnt in its list of font characters, in which case it gives up and uses MeasureText() on the message instead. as long as your list of characters covers 99.9% of speech it only falls back in the tiniest of edge cases

File: code/controllers/subsystem/runechat.dm
Comment (Kylerace) [inline code/controllers/subsystem/runechat.dm:102]: if we're at all worried about bad actors like lemon was talking about last night, it would be better to randomize which client is picked in each of these steps

File: code/datums/chatmessage.dm
Comment (Kylerace) [reply code/datums/chatmessage.dm:50]: they have been unused since SSrunechat was refactored to be a subtype of SStimer

**Review by Kylerace — CHANGES_REQUESTED** [review]

File: code/datums/chatmessage.dm
Comment (Tastyfish) [reply code/datums/chatmessage.dm:339]: Small Fonts *does* have the accented characters for the major european languages. It would be nice if it at least supported those.

image

Comment (Tastyfish) [reply code/datums/chatmessage.dm:339]: Also, it appears that this could be a global function since it doesn't use src at all as far as I can see, and then `balloon_alert_perform()` could also use it.

Or possibly also to make status displays seemlessly loop. Status displays only have the gaps after 3 repetitions because it was easier to do that than actually attempt to measure text when there's no users for them.

File: code/controllers/subsystem/runechat.dm
Comment (Mothblocks) [reply code/controllers/subsystem/runechat.dm:102]: Isn't it a much better and consistent idea to just cache per client?

File: code/datums/chatmessage.dm
Comment (Archanial) [reply code/datums/chatmessage.dm:339]: On bee balloon alerts are really subtypes of chatmessages.
I could either repath it and implement the function there.
Or make the proc global and reuse it. 
I'll leave it up to maintainers if they want either.

Comment (Mothblocks) [reply code/datums/chatmessage.dm:339]: Do not make them a subtype of chat messages, make the proc global

File: code/datums/chatmessage.dm
Comment (Archanial) [reply code/datums/chatmessage.dm]: Removed

File: code/datums/chatmessage.dm
Comment (Archanial) [reply code/datums/chatmessage.dm]: Done

**Comment by Archanial** [comment]
I've implemented the per client cache.
It's being currently testmerged on bee and first thing you can notice is that the values are universal per client.
With that, maybe it would be possible to cache it globally again, but instead take, say, 3 random clients, so that one bad client wouldn't ruin it for all.

**Comment by Tastyfish** [comment]
> I've implemented the per client cache. It's being currently testmerged on bee and first thing you can notice is that the values are universal per client. With that, maybe it would be possible to cache it globally again, but instead take, say, 3 random clients, so that one bad client wouldn't ruin it for all.

Yeah, polling 3 clients and determining they all yield consensus should be fine IHMO.

**Comment by Archanial** [comment]
It should add them to cache after the first use.

also this isn't dead Im just lazy and busy

**Comment by Mothblocks** [comment]
This has staled out twice, let me know when you're working on this again so I can unclose

**Comment by LemonInTheDark** [comment]
I hunger for this

**Comment by Archanial** [comment]
I got kinda stuck figuring out how to implement the cache.
Per player is kinda meh, since it results in lots of same data, plus it's really hard to controll whether this player has a complete cache and cache letter globally.
Ideally I would just instanciate a client and take results from a it, since it can be trusted, but it's also not possible.
I wonder if you can run measure text on the server itself.

**Comment by Mothblocks** [comment]
The MeasureText proc itself no because there's nothing close to a renderer

**Comment by Kylerace** [comment]
oranges_scribe when

**Comment by Archanial** [comment]
Another idea would be to run it on the tgui, you could then measure server side.

**Comment by Mothblocks** [comment]
that has the exact same client trust issues

**Pull request closed without merging**
